src/VisualStudio/Core/Def/Implementation/Diagnostics/VisualStudioWorkspaceDiagnosticAnalyzerProviderService.cs

```diff
@@ -46,6 +71,56 @@ public IEnumerable<HostDiagnosticAnalyzerPackage> GetHostDiagnosticAnalyzerPacka
             return _hostDiagnosticAnalyzerInfo;
         }
 
+        private bool TryGetRootAndShellFolder(IVsExtensionManager extensionManager, out string shellFolder, out string rootFolder)
+        {
+            // use reflection to get this information. currently there is no other way to get this information
+            shellFolder = GetProperty(extensionManager, "ShellFolder");
+            rootFolder = GetProperty(extensionManager, "RootFolder");
+
+            return !string.IsNullOrEmpty(rootFolder) && !string.IsNullOrEmpty(shellFolder);
+        }
+
+        private string GetContentLocation(string shellFolder, string rootFolder, string installPath, string relativePath)
+        {
+            // extension manager should expose an API that doesnt require this.
+            const string ShellFolderToken = "$ShellFolder$";
+            const string RootFolderToken = "$RootFolder$";
+
+            if (relativePath.StartsWith(ShellFolderToken))
+            {
+                return relativePath.Replace(ShellFolderToken, shellFolder);
+            }
+            else if (relativePath.StartsWith(RootFolderToken))
+            {
+                return relativePath.Replace(RootFolderToken, rootFolder);
+            }
+
+            string contentLocation = null;
+            try
+            {
+                contentLocation = Path.Combine(installPath, relativePath);
+            }
+            //Path.Combine will throw an ArgumentException if either of the two path arguments contain illegal characters.
+            //We'll just catch this exception here and ignore the paths with illegal characters.
+            catch (ArgumentException)
+            {
+            }
+
+            return contentLocation;
```

Comment:
Should we also check that contentLocation points to a valid file on disk just to be sure?